src/ratbag_emu/protocol/base.py

```diff
@@ -201,31 +201,37 @@ def _send_packets(self, packets, total):
 
 
 class BaseDevice(object):
-    def __init__(self, rdesc=None, info=None, name='Generic Device',
-                 shortname='generic'):
-        self.info = info
-        self.name = name
-        self.shortname = shortname
+    def __init__(self, obj):
+        self.info = tuple(obj['bus_info'])
+        self.name = obj['name']
+        self.shortname = obj['shortname']
         self.endpoints = {}
 
-        self.endpoints[0] = SimulatedDevice(self, rdesc, info, name, shortname)
+        for i, rdesc in obj['report_descriptors'].items():
+            self.endpoints[i] = Endpoint(self, rdesc)
 
-        self.protocol = None
+        assert len(self.endpoints), 'At least one report_descriptor is required'
 
-
-        # Represents the active profile in the hardware.
-        self.hw_profile = Profile()
-
-        self.active_profile = None
         self.profiles = []
 
+        self.active_profile = None
+        if 'profiles' in obj:
+            for i, profile in enumerate(obj['profiles']):
+                self.profiles.append(Profile(profile))
+                if profile['is_active']:
+                    self.active_profile = i
```

Comment:
You are using `enumerate` so you can store the index of the active profile. I'd use a simple for/in loop and keep a reference to the active profile instead:
```python
 for profile in obj['profiles']:
    p = Profile(profile)
    self.profiles.append(p)
    if profile['is_active']:
          self.active_profile = p
```

Reply:
The main reason behind the existence of an `self.active_profile` field is not to store the active profile, but an index of the active profile as some protocols will need to export that information. The real active profile, which I call hardware profile is stored in `self.hw_profile`. If you really want, I can change this but just keep in mind I am opposed to it.

Reply:
> The main reason behind the existence of an self.active_profile field is not to store the active profile, but an index of the active profile as some protocols will need to export that information.

I am not asking you to store the entire profile, but a reference to it. And if you *really* need the index, you can either recompute it or store the index in the profile itself (but a property would be better IMO).

> The real active profile, which I call hardware profile is stored in self.hw_profile. If you really want, I can change this but just keep in mind I am somewhat opposed to it.

I honestly never understood the concept of the `self.hw_profile` vs `self.active_profile`. There is a difference that I do not get. And if the difference is because some random mouse has one, then we should have the difference in the protocol, not in the generic base class.

So one more thing to do: remove the `self.hw_profile` and convert it to `self.active_profile`. In a separate commit, please.

Reply:
> I am not asking you to store the entire profile, but a reference to it. And if you really need the index, you can either recompute it or store the index in the profile itself (but a property would be better IMO).

```python
self.hw_profile = copy.deepcopy(self.profiles[self.active_profile])
```

This is the only line in the code that will make use of it. My motivation was that we don't really need compute the index each time we need it. But I guess that is just a nitpick. I will change it as you requested.

> I honestly never understood the concept of the self.hw_profile vs self.active_profile. There is a difference that I do not get. And if the difference is because some random mouse has one, then we should have the difference in the protocol, not in the generic base class.
>
> So one more thing to do: remove the self.hw_profile and convert it to self.active_profile. In a separate commit, please.

Several protocols allow you to change the RAM settings without touching the ROM. This means the active settings on the mouse (RAM) might not match the active settings on the profile (ROM). Steelseries does it, Logitech does it and I am pretty sure Razer and Corsair do it too.

Do you need me to elaborate more? Do you still think we should remote it? Let me know.

Reply:
I think you are mixing a few concepts here:
- ROM: Read Only Memory -> memory that is write once in the factory and can never be overwritten
- RAM: Random Access Memory -> volatile memory on a chip
- Flash -> high latency memory, but persistent, used to store the profiles.

(I get that you used ROM for Flash)

Anyway, the problem here is we do not want the emulator to replicate what subtleties a device can have but we want an ideal one.

Here, you are talking about the mice that can be put in the software vs hardware control. Software controlled mice are generally using a RAM profile, as there is no point in storing it on the flash given that the software on the host controls it.

But here, it *really* doesn't matter. It's a firmware implementation that needs to be dealt with in the various protocols.

Our ideal device needs to have several profiles and one of them is the active profile. The latency is null, and we do not care whether we are using software or hardware control. We want to have one active profile we can change, and several other profiles we can swap.

But when you deal with Logitech, Steelseries, or Razer, then yes, you need to implement at most all of the subtleties of the firmware. So with a Logitech mouse in the software mode, a protocol request to change the current dpi would call `set_dpi()` on `BaseDevice`, which would change the currently active profile on the emulator. But when you get a request to read the flash, then you should return whatever you kept in the flash (in `HIDPP20Device`), not the active profile.

Reply:
Yes, my bad. I meant Flash, not ROM.

> But here, it really doesn't matter. It's a firmware implementation that needs to be dealt with in the various protocols.

I was trying to ease the implementation of those features in the specific protocol classes as it might be a bit tricky, and we want people to contribute to the project. Having support for this in the backend would be mice IMO but I guess I can change it.

Reply:
I think "trying to ease the implementation of those features in the specific protocol classes" is taking the problem backward. We should have a simple and reliable mouse/keyboard/device simulator that does one thing only: emulate events.

We are already handling profiles in this simulator, and I am more and more thinking it should probably be handled in a separate protocol (the generic one). So if the `BaseDevice` class has only the current state in memory and no other state, it would be the responsibility of the protocol to handle profiles, software profiles, hardware profiles and the rest. There might be code duplication in the protocol to handle the profiles, but we can deal with that later.

Furthermore, if I were to write the HID++2.0 hardware profiles feature, I do not want to sync my state with the emulator, and I'd be actually happier to not have to do it. Whenever I receive a get profiles command, I would just return my flash and forget about the rest.

Reply:
> and I am more and more thinking it should probably be handled in a separate protocol

This is perfectly fine by me. We can have ratbagd-like profiles in the a `GenericDevice`. `BaseDevice` would just have a `self.profile` field.